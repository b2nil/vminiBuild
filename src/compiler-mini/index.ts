export { compileScript } from "./script"
export { compileTemplate, normalizeAssetsURLOptions, getPlatformDirective } from "./template"
export { parse, compileStyleAsync } from "@vue/compiler-sfc"