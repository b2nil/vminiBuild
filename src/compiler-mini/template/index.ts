import { stringifyNodes } from './stringify'
import { transformSyntax, transformAssetUrls, transformStyleWithCssVars, transformCssModuleClasses } from './transforms'
import { baseParse, CompilerError, transform } from "@vue/compiler-core"

import type { VueOptions } from 'types'
import type { ElementNode } from '@vue/compiler-core'
import type {
  SFCTemplateCompileOptions,
  SFCTemplateCompileResults
} from '@vue/compiler-sfc'

interface PreProcessor {
  render (
    source: string,
    options: any,
    cb: (err: Error | null, res: string) => void
  ): void
}

function preprocess (
  preprocessor: PreProcessor,
  {
    source,
    filename,
    preprocessOptions
  }: Pick<SFCTemplateCompileOptions,
    "source" | "filename" | "preprocessOptions">
): string {
  let res: string = ''
  let err: Error | null = null

  preprocessor.render(
    source,
    { filename, ...preprocessOptions },
    (_err, _res) => {
      if (_err) err = _err
      res = _res
    }
  )

  if (err) throw err
  return res
}

export { getPlatformDirective } from "./utils"
export { normalizeAssetsURLOptions } from "./transforms"

export async function compileTemplate (
  template: string,
  options: VueOptions["template"] & {
    filename: string,
    id: string,
    cssVars: string[],
    isComponent?: boolean,
    cssModules: Record<string, Record<string, string>>,
    platformDir: string
  }
): Promise<SFCTemplateCompileResults> {
  const errors: CompilerError[] = []
  const warnings: CompilerError[] = []
  const preprocessLang = options?.preprocessLang

  if (preprocessLang && preprocessLang === "pug") {
    try {
      const pug = options.preprocessCustomRequire!("pug")
      template = preprocess(pug, {
        source: template,
        filename: options.filename,
        preprocessOptions: options.preprocessOptions
      })
      // Fix #default="#default" and v-else="v-else"
      template = template.replace(/(\B#.*?|\bv-.*?)="\1"/g, "$1")
    } catch (e: any) {
      return {
        code: `export default function render() {}`,
        source: template,
        tips: [],
        errors: [e]
      }
    }
  }

  const assetOptions = options.transformAssetUrls
    ? typeof options.transformAssetUrls === "boolean"
      ? {}
      : options.transformAssetUrls
    : {}

  const longId = `data-v-${options.id}`

  const ast = baseParse(template, options?.compilerOptions || {})
  transform(ast, {
    ...(options?.compilerOptions || {}),
    scopeId: options.scoped ? longId : undefined,
    nodeTransforms: [
      ...(options?.compilerOptions?.nodeTransforms || []),
      transformCssModuleClasses(options.cssModules),
      transformSyntax(options.platformDir),
      transformAssetUrls(assetOptions),
      transformStyleWithCssVars(options.id, options.cssVars, options.isComponent),
    ]
  })

  const code = stringifyNodes(ast.children, ast as unknown as ElementNode, options.platformDir)

  return {
    code,
    ast,
    source: template,
    tips: [],
    errors
  }
}