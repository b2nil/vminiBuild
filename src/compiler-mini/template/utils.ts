const platformDirMap: Record<string, string> = {
  weapp: "wx",
  swan: "s",
  alipay: "a",
  tt: "tt",
  jd: "jd",
}

export const __DIR = platformDirMap[process.env.__PLATFORM__!]

export const eventMap: Record<string, string> = {
  click: "tap"
}

export const directivesMap: Record<string, string> = {
  if: `${__DIR}:if`,
  "else-if": `${__DIR}:elif`,
  else: `${__DIR}:else`,
  show: "hidden"
}

export const eventModifiers: Record<string, string> = {
  stop: "catch", // .stop => catch
  catch: "catch", // .catch => catch
  mut: "mut-bind",  // .mut => mut-bind
  capture: "capture-bind", // .capture => capture-bind
}

export function getEventModifier (modifiers: string[]) {
  for (const m of modifiers) {
    if (!eventModifiers[m]) {
      console.warn(`Modifier .${m} not supported yet.`)
      console.warn(`Currently supporting: .stop/.catch, .mut, .capture, .capture.catch`)
    }
  }

  if (modifiers.length === 1 &&
    eventModifiers[modifiers[0]]
  ) {
    return eventModifiers[modifiers[0]]
  }

  // .capture.stop or .capture.catch => capture-catch
  if (modifiers.includes("capture") &&
    (modifiers.includes("stop") || modifiers.includes("catch"))
  ) {
    return "capture-catch"
  }

  return "bind"
}


