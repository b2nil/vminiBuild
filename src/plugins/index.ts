export * from "./cache"
export * from "./vue"
export * from "./style"
export * from "./entry"
export { printStats } from "./emits"