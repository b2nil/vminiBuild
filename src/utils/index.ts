export * from "./cache"
export * from "./paths"
export * from "./config"
export * from "./utils"

export const enum DEFINE {
  PROPS = "defineProps",
  EXPOSE = "defineExpose",
  HOOK = "defineHookConfig",
  APP_CONFIG = "defineAppConfig",
  PAGE_CONFIG = "definePageConfig",
  BUILD_CONFIG = "defineBuildConfig",
  PROJECT_CONFIG = "defineProjectConfig",
  SITEMAP_CONFIG = "defineSitemapConfig",
}