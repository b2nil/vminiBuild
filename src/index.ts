import fs from "fs"
import path from 'path'
import esbuild from "esbuild"
import { cac } from "cac"
import { customRequire, DEFINE, extractConfigFromFile } from './utils'
import {
  printStats,
  vuePlugin,
  styleLoader,
  nativePlugin,
  vueminiPlugin,
} from "./plugins"

import type { BuildOptions } from "esbuild"
import type { UserConfig, CliOptions } from 'types'
import { normalizeAssetsURLOptions } from './compiler-mini'

const cli = cac("vmini")

async function getUserBuildConfig (opts: CliOptions): Promise<UserConfig> {
  let configFile = opts.c || opts.config
  let resolvedPath: string | undefined

  if (configFile) {
    const specifiedPath = path.resolve(process.cwd(), configFile)
    if (fs.existsSync(specifiedPath)) {
      resolvedPath = specifiedPath
    }
  } else {
    const tsconfigFile = path.resolve(process.cwd(), `mini.config.ts`)
    if (fs.existsSync(tsconfigFile)) {
      resolvedPath = tsconfigFile
    }

    if (!resolvedPath) {
      const jsconfigFile = path.resolve(process.cwd(), `mini.config.js`)
      if (fs.existsSync(jsconfigFile)) {
        resolvedPath = jsconfigFile
      }
    }
  }

  const buildConfig: UserConfig = {
    outDir: `dist`,
    emptyOutDir: false,
    platform: "weapp",
    designWidth: 750,
    logLevel: "debug"
  }

  if (!resolvedPath) {
    console.warn(`config file mini.config.ts/js not found`)
    console.warn(`use default config`)
    return buildConfig
  }

  const config = await extractConfigFromFile(resolvedPath, DEFINE.BUILD_CONFIG)

  if (!config) {
    console.warn(`Failed to read build config at: ${resolvedPath}`)
    console.warn("Please use `export default` or `exports.default` to export the build config")
    console.warn("Now use the build-in config instead.")
    return buildConfig
  }

  return config as UserConfig
}

async function initBuildOptions (userConfig: UserConfig): Promise<BuildOptions> {
  const { useCDN, ...userBuildOptions } = userConfig
  userBuildOptions.vue = {
    ...(userBuildOptions.vue || {}),
    useCDN,
    style: {
      ...(userBuildOptions.vue?.style || {}),
      postcssPlugins:
        userBuildOptions.vue?.style?.postcssPlugins
          ? userBuildOptions.vue?.style?.postcssPlugins
          : [
            customRequire(`postcss-pxtransform`)({
              platform: userBuildOptions.platform || `weapp`,
              designWidth: userBuildOptions.designWidth || 750
            })
          ],
      preprocessCustomRequire: customRequire
    },
    template: {
      ...(userBuildOptions.vue?.template || {}),
      preprocessCustomRequire: customRequire
    }
  }

  if (useCDN) {
    let { transformAssetUrls } = userBuildOptions.vue.template!

    if (!transformAssetUrls || transformAssetUrls === true) {
      transformAssetUrls = {
        base: `http://${useCDN.host}:${useCDN.port}/`
      }
    } else {
      transformAssetUrls = normalizeAssetsURLOptions(transformAssetUrls, {
        base: `http://${useCDN.host}:${useCDN.port}/`,
        includeAbsolute: false,
        tags: {}
      })
    }

    userBuildOptions.vue.template!["transformAssetUrls"] = transformAssetUrls
  }


  const buildOptions: BuildOptions = {
    entryPoints: ["src/app.ts", "src/app.config.ts"],
    outdir: userBuildOptions.outDir || `dist`,
    target: "esnext",
    format: "esm",
    bundle: true,
    color: true,
    metafile: true,
    incremental: true,
    splitting: true,
    define: {
      ...(userBuildOptions.define || {}),
      "process.env.__PLATFORM__": JSON.stringify(userBuildOptions.platform || "weapp")
    },
    watch: !userBuildOptions.watch ? false : {
      onRebuild (error, _res) {
        if (error) console.error('[x] rebuild failed:', error.message, "\n", error.stack)
        else console.log('[+] rebuild succeeded')
      }
    },
    logLevel: userBuildOptions.logLevel,
    chunkNames: 'common/[dir]/[name]-[hash]',
    assetNames: 'assets/[dir]/[name]-[hash]',
    plugins: [
      vueminiPlugin({
        watch: userBuildOptions.watch,
        minify: userBuildOptions.minify,
        aliases: userBuildOptions.aliases
      }),
      vuePlugin(userBuildOptions.vue),
      styleLoader(userBuildOptions.vue),
      nativePlugin(userBuildOptions.vue),
      ...(userBuildOptions.plugins || [])
    ]
  }

  return buildOptions
}

async function buildWithEsbuild (options: CliOptions) {
  const userConfig = await getUserBuildConfig(options)

  if (options.p || options.platform) {
    userConfig.platform = options.p || options.platform
  }

  process.env.__PLATFORM__ = userConfig.platform || 'weapp'

  userConfig.watch = options.w || options.watch
  userConfig.logLevel = options.l || options.logLevel
  userConfig.emptyOutDir = options.emptyOutDir || userConfig.emptyOutDir
  userConfig.minify = options.minify || userConfig.minify

  const buildOptions = await initBuildOptions(userConfig)

  // temprary hack to serve the whole src dir
  if (userConfig.useCDN) {
    esbuild.serve({
      ...userConfig.useCDN,
      servedir: `src`
    }, {
      entryPoints: [],
      outdir: `src/${userConfig.useCDN?.servedir || "images"}`,
      write: false
    }).then(_server => {
      // do nothing
    })
  }

  esbuild.build(buildOptions)
    .then(async (res) => {
      if (buildOptions.watch || userConfig.useCDN) {
        if (userConfig.useCDN) {
          console.log(`[+] serving static files at: http://${userConfig.useCDN.host}:${userConfig.useCDN.port}`)
        }
        console.log('[+] watching...')
      } else {
        await printStats(res.metafile!, esbuild)
      }
    })
    .catch((e: any) => {
      console.error(`[x] build error: \n${e.stack || e}`)
    })
}

export function printVersion () {
  const root = path.resolve(__dirname, "../")
  const ver = require(path.join(root, "package.json")).version
  console.log(`(^o^) vminiBuild v${ver} (^o^)`)
}

export function run () {
  cli
    .option("-p, --platform [platform]", "[string] weapp | alipay | swan | tt", { default: "weapp" })
    .option("-l, --loglevel <level>", `[string] debug | warn`)
    .option("-c, --config <configFilePath>", `[string] specify config file path, defualt: mini.config.ts/js`)

  cli
    .command("build")
    .option("-w, --watch", "[boolean] enable watch mode")
    .option("--minify", "[boolean] force minify the outputs")
    .option("--emptyOutDir", "[boolean] force empty OutDir")
    .action(buildWithEsbuild)

  cli.help()
  cli.parse()
}
