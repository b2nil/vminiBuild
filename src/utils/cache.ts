import { normalizePath } from './utils'

export const assetsCache = new Set<string>([])
export const moduleImportsCache = new Set<string>([])
export const utilsImportsCache = new Set<string>([])

export function cacheAssetPath (assetURL: string) {
  cachePath(assetURL, assetsCache)
}

export function cacheUtilsImports (assetURL: string) {
  cachePath(assetURL, utilsImportsCache)
}

export function cacheModuleImports (assetURL: string) {
  cachePath(assetURL, moduleImportsCache)
}

function cachePath (p: string, cache: Set<string>) {
  const normalized = normalizePath(p)
  if (!cache.has(normalized)) {
    cache.add(normalized)
  }
}