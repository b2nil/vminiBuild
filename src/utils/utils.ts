import os from "os"
import path from "path"
import fs from "fs"
import { replaceRules } from './paths'
import { moduleImportsCache, utilsImportsCache } from './cache'

import type { OnResolveArgs, PluginBuild } from "esbuild"

export enum EXTENSIONS {
  WXML = ".wxml",
  JSON = ".json",
  WXSS = ".wxss",
  JS = ".js",
  WXS = ".wxs"
}

export const __OUT__ = { dir: `dist` }
export const __PROD__ = process.env.NODE_ENV === "production"
export const appREG = /app\.(t|j)sx?$/
export const appConfigREG = /app\.config\.(t|j)sx?$/
export const cssREG = /\.(css|less|scss|stylus|sass)$/
export const styleExts = ["css", "less", "scss", "stylus", "sass"]
export const customRequire = (id: string) => require(path.resolve(process.cwd(), `node_modules/${id}`))

export const isTS = (f: string) => path.extname(f) === ".ts"
export const isWindows = os.platform() === 'win32'
export function normalizePath (id: string): string {
  return isWindows ? id.replace(/\\/g, "\\\\") : id
}

export const getOutputFilename = (
  p: string,
  ext: string
) => p.replace("src", __OUT__.dir).replace(".vue", ext)

export async function emitFile (filename: string, ext: string, code: string) {
  const outputFilename = getOutputFilename(filename, ext)
  const dir = path.dirname(outputFilename)

  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true })
  }

  fs.promises.writeFile(outputFilename, code, "utf8")
}

function getFilename (p: string) {
  const exts = [".ts", "/index.ts", ".js", "/index.js"]
  for (const e of exts) {
    const temp = e.startsWith(".") ? p + e : path.join(p, e)
    if (fs.existsSync(temp)) {
      p = temp
      break
    }
  }
  return p
}

export function resolveImports (namespace: string, build: PluginBuild) {
  return build.onResolve({ filter: /.*/, namespace }, async (args) => {
    if (
      (/^(\.|\@\/)/.test(args.path) && !args.path.endsWith(".vue")) ||
      externals.includes(args.path)
    ) {
      args.path = replaceRules(args.path)
      if (externals.includes(args.path)) {
        if (!moduleImportsCache.has(args.path)) {
          moduleImportsCache.add(args.path)
        }

        return {
          path: args.path,
          external: true
        }
      }

      const fullpath = normalizePath(getFilename(getFullPath(args)))

      if (!utilsImportsCache.has(fullpath)) {
        utilsImportsCache.add(fullpath)
      }

      return {
        path: args.path,
        external: true
      }
    }
  })
}

export function parseVueRequest (search: string): Record<string, string> {
  let hashes = search.slice(search.indexOf('?') + 1).split('&')
  return hashes.reduce((params, hash) => {
    let [key, val] = hash.split('=')
    return Object.assign(params, { [key]: decodeURIComponent(val) })
  }, {})
}

export function getFullPath (args: OnResolveArgs) {
  return path.isAbsolute(args.path) ? args.path : path.join(args.resolveDir, args.path)
}

const getDeps = (): string[] => {
  const pkgPath = normalizePath(path.join(process.cwd(), `package.json`))
  const pkg = require(pkgPath)
  return Object.keys(pkg.dependencies)
}

export const externals = getDeps()

export function checkFileExists (mod: string): string {
  const exts = [".ts", ".js"]
  let filename = mod
  for (const e of exts) {
    let temp = mod + e
    if (fs.existsSync(temp)) {
      filename = temp
      break
    } else if (fs.existsSync(path.join(mod, "index" + e))) {
      filename = path.join(mod, "index" + e)
      break
    }
  }
  return filename
}