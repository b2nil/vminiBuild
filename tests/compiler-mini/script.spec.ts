import path from "path"
import { compileScript } from '../../src/compiler-mini'
import { externals } from '../../src/utils'
import { parse } from '@vue/compiler-sfc'
import type { TransformResult } from '../../types'

externals.push("@vue-mini/wechat")

const script = `
<script>
import { ref } from "@vue-mini/wechat"
import ComOne from "../components/comp.vue"
import ComTwo from "../components/comp2.vue"
definePageConfig({
  a: {c: "d"},
  b: "a"
})
export default definePage({
  components: {
    ComOne,
    ComTwo
  },
  props: { hi: Number },
  setup(){
    const data = ref(0)
    const method = () => { console.log() }
    return { 
      data,
      method
    }
  }
})
</script>
`

const scriptSetup = `
<script setup>
import { ref } from "@vue-mini/wechat"
import ComOne from "../components/comp.vue"
import ComTwo from "../components/comp2.vue"

definePageConfig({
  a: {c: "d"},
  b: "a"
})

defineProps({
  hi: Number
})

defineHookConfig({
  listenPageScroll: true
})

const data = ref(null)

const method = () => { console.log() }

defineExpose({
  data,
  method
})
</script>
`
const extractedConfig = {
  a: { c: "d" },
  b: "a",
  usingComponents: {
    "com-one": "../components/comp.vue",
    "com-two": "../components/comp2.vue"
  }
}

describe("transform script", () => {
  it("should transform normal script", async () => {
    const { descriptor } = parse(script, {
      filename: "./test.vue"
    })
    let ret: TransformResult = {}
    ret = compileScript(descriptor, ret)

    expect(ret.config).toEqual(extractedConfig)

    expect(ret.components).toEqual({
      "com-one": "../components/comp.vue",
      "com-two": "../components/comp2.vue"
    })

    expect(ret.imports).toEqual(new Set<string>([
      `${path.resolve(process.cwd(), "../components/comp.vue")}`,
      `${path.resolve(process.cwd(), "../components/comp2.vue")}`,
    ]))

    expect(ret.code).not.toMatch(/"\.\.\/components\/.+\.vue"/)
    expect(ret.code).toMatchSnapshot()

    const post = compileScript(descriptor, ret, undefined, true)
    expect(post.code).toMatchSnapshot()
    expect(ret.config).toMatchSnapshot()
  })

  it("should transform scriptSetup", async () => {
    const { descriptor } = parse(scriptSetup, {
      filename: "./test.vue"
    })
    let ret: TransformResult = {}
    ret = compileScript(descriptor, ret)

    expect(ret.config).toEqual(extractedConfig)

    expect(ret.components).toEqual({
      "com-one": "../components/comp.vue",
      "com-two": "../components/comp2.vue"
    })

    expect(ret.imports).toEqual(new Set<string>([
      `${path.resolve(process.cwd(), "../components/comp.vue")}`,
      `${path.resolve(process.cwd(), "../components/comp2.vue")}`,
    ]))
    expect(ret.code).toMatchSnapshot()

    const post = compileScript(descriptor, ret, undefined, true)
    expect(post.code).toMatchSnapshot()
    expect(post.config).toMatchSnapshot()
  })
})