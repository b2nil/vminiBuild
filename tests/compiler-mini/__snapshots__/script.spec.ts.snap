// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform script should transform normal script 1`] = `
"
import { ref } from \\"@vue-mini/wechat\\"
import ComOne from \\"../components/comp.vue\\"
import ComTwo from \\"../components/comp2.vue\\"
definePageConfig({
  a: {c: \\"d\\"},
  b: \\"a\\"
})
export default definePage({
  components: {
    ComOne,
    ComTwo
  },
  props: { hi: Number },
  setup(){
    const data = ref(0)
    const method = () => { console.log() }
    return { 
      data,
      method
    }
  }
})

function definePageConfig(config) { return config }
"
`;

exports[`transform script should transform normal script 2`] = `
"import { ref } from \\"@vue-mini/wechat\\";
export default definePage({
  properties: { hi: Number },
  setup() {
    const data = ref(0);
    const method = () => {
      console.log();
    };
    return {
      data,
      method
    };
  }
});
"
`;

exports[`transform script should transform normal script 3`] = `
Object {
  "a": Object {
    "c": "d",
  },
  "b": "a",
  "usingComponents": Object {
    "com-one": "../components/comp",
    "com-two": "../components/comp2",
  },
}
`;

exports[`transform script should transform scriptSetup 1`] = `
"
import { ref } from \\"@vue-mini/wechat\\"
import ComOne from \\"../components/comp.vue\\"
import ComTwo from \\"../components/comp2.vue\\"

definePageConfig({
  a: {c: \\"d\\"},
  b: \\"a\\"
})

defineProps({
  hi: Number
})

defineHookConfig({
  listenPageScroll: true
})

const data = ref(null)

const method = () => { console.log() }

defineExpose({
  data,
  method
})

function definePageConfig(config) { return config }
function defineProps(config) { return config }
function defineExpose(config) { return config }
function defineHookConfig(config) { return config }
const COMP_r3m0v3 = { ComOne, ComTwo }
export default COMP_r3m0v3
"
`;

exports[`transform script should transform scriptSetup 2`] = `
"import { ref, defineComponent } from \\"@vue-mini/wechat\\";
export default defineComponent({
  properties: {
    hi: Number
  },
  setup() {
    const data = ref(null);
    const method = () => {
      console.log();
    };
    return {
      data,
      method
    };
  }
}, {
  listenPageScroll: true
});
"
`;

exports[`transform script should transform scriptSetup 3`] = `
Object {
  "a": Object {
    "c": "d",
  },
  "b": "a",
  "usingComponents": Object {
    "com-one": "../components/comp",
    "com-two": "../components/comp2",
  },
}
`;
